{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_orig_metadata(metadata_path='../data/barcodes_BMC_orig_ds1.xlsx', \n",
    "                        outdir='../data/metadata_and_manifest/Blactam_metadata_ds1.tsv', \n",
    "                        save=True):\n",
    "    \n",
    "    \"\"\"\n",
    "    parse the original metadata file provided by BioMicro Center (https://openwetware.org/wiki/BioMicroCenter) \n",
    "    such that it is comptatible with the qiime2 metadata format. \n",
    "    this file will be exported as a tsv which can be uploaded to\n",
    "    Google Sheets for additional validation via Keemei (https://keemei.qiime2.org/)\n",
    "    \n",
    "    parameters\n",
    "        metadata_path: string\n",
    "            path to original metadata file\n",
    "        outdir: string\n",
    "            directory and filename in which to output metadata file\n",
    "        save: boolean\n",
    "            indicate whether to save tsv file to specified directory\n",
    "    returns:\n",
    "        metadata_df: pandas dataframe\n",
    "            dataframe containing metadata in qiime2-compatible format\n",
    "    \"\"\"\n",
    "    \n",
    "    # parse through 180 samples\n",
    "    metadata_bmc_orig_df = pd.read_excel(metadata_path).loc[0:179,:]\n",
    "    \n",
    "    # list in which to store metadata\n",
    "    metadata_bmc_parsed_list = []\n",
    "    # specify variable types\n",
    "    metadata_bmc_parsed_list.append({\n",
    "        'sample-id': '#q2:types',\n",
    "        'barcode-sequence': 'categorical',\n",
    "        'treatment-group': 'categorical',\n",
    "        'time-point-day': 'categorical',\n",
    "        'animal-number': 'categorical',\n",
    "        'treatment-group-time-mouse': 'categorical',\n",
    "        'time-treatment-group-mouse': 'categorical',\n",
    "        'treatment-group-time': 'categorical'     \n",
    "    })\n",
    "    \n",
    "    # create dictionaries that map treatment day and animal numbers to letters -- this will help organize taxa barplots\n",
    "    alphabet_string = string.ascii_lowercase\n",
    "    alphabet_list = list(alphabet_string)\n",
    "    alphabet_list = [letter.upper() for letter in alphabet_list]\n",
    "    alphabet_list_trun = alphabet_list[0:10]\n",
    "    alphabet_list_trun.reverse()\n",
    "    alphabet_list_trun = alphabet_list_trun + ['J', 'I', 'H']\n",
    "\n",
    "    treatment_days = [0, 1, 2, 3, 5, 7, 9, 12, 'L1', 'L2', 'L3']\n",
    "    animal_numbers = list(range(1,21))\n",
    "    treatment_days_to_letter_dict = dict(zip(treatment_days,alphabet_list_trun))\n",
    "    animal_numbers_to_letter_dict = dict(zip(animal_numbers,alphabet_list_trun[0:10] +  alphabet_list_trun[0:10]))\n",
    "\n",
    "    # parse every row for metadata\n",
    "    for index, row in metadata_bmc_orig_df.iterrows():\n",
    "        \n",
    "        # parsing information from first column (should contain sample-id and barcode)\n",
    "        split_sample_name = row['Unnamed: 0'].replace('_','-').replace('\\xa0','').split(':')\n",
    "        sample_id = split_sample_name[1].split(' ')[1]\n",
    "        sample_barcode = split_sample_name[1].replace('(','').replace(')','').split(' ')[-1]\n",
    "\n",
    "        # extracting information from other columns\n",
    "        treatment_group = row['Treatment Group']\n",
    "        time_point_day = row['Time Point (Day)']\n",
    "        animal_number = row['Animal Number']\n",
    "        \n",
    "        treatment_group_time_mouse = treatment_group + '-' + treatment_days_to_letter_dict[time_point_day] \\\n",
    "        + 'day' + str(time_point_day) + '-' + animal_numbers_to_letter_dict[animal_number] + 'animal' + str(animal_number) \n",
    "        \n",
    "        time_treatment_group_mouse = treatment_days_to_letter_dict[time_point_day] \\\n",
    "        + 'day' + str(time_point_day) + '-' + treatment_group \\\n",
    "        + '-' + animal_numbers_to_letter_dict[animal_number] + 'animal' + str(animal_number) \n",
    "        \n",
    "        treatment_group_time = treatment_group + '-' + 'day' + str(time_point_day)\n",
    "        \n",
    "        # update list with metadata\n",
    "        metadata_bmc_parsed_list.append({\n",
    "        'sample-id': sample_id,\n",
    "        'barcode-sequence': sample_barcode,\n",
    "        'treatment-group': treatment_group,\n",
    "        'time-point-day': time_point_day,\n",
    "        'animal-number': animal_number,\n",
    "        'treatment-group-time-mouse': treatment_group_time_mouse,\n",
    "        'time-treatment-group-mouse': time_treatment_group_mouse,\n",
    "        'treatment-group-time': treatment_group_time     \n",
    "    })\n",
    "        \n",
    "    metadata_df = pd.DataFrame(metadata_bmc_parsed_list)\n",
    "    \n",
    "    if save:\n",
    "        metadata_df.to_csv(outdir, sep='\\t', index=False)\n",
    "    return(metadata_df)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_manifest_file(metadata_df, \n",
    "                         outdir='../data/metadata_and_manifest/Blactam_manifest_ds1.tsv', \n",
    "                         save=True):\n",
    "    \n",
    "    \"\"\"\n",
    "    create a manifest file detailing the location of all sequencing reads.\n",
    "    the manifest file will be exported as a tsv file\n",
    "    \n",
    "    parameters\n",
    "        metadata_df: pandas dataframe\n",
    "            metadata dataframe that contains sample-ids\n",
    "        outdir: string\n",
    "            directory and filename in which to output manifest file\n",
    "        save: boolean\n",
    "            indicate whether to save tsv file to specified directory\n",
    "    returns\n",
    "        manifest_df: pandas dataframe\n",
    "            manifest file detailing the directory path of sequencing data\n",
    "        \n",
    "    \"\"\"\n",
    "    \n",
    "    sample_numbers = [\"%.3d\" % i for i in range(1,181)]\n",
    "    \n",
    "    # path where sequencing data will be located \n",
    "    seq_path_prefix = '$PWD/../data/Blactam_16s_data/210616ChiA/'\n",
    "    manifest_file_list = []\n",
    "\n",
    "    for samp_num, sample_id in zip(sample_numbers, list(metadata_df.iloc[1:,0])):\n",
    "\n",
    "        samp_num_full = 'D21-217' + samp_num \n",
    "        \n",
    "        # data run on two separate lanes, so need to consider differing file suffix\n",
    "        # sample 001 through 090 were run on 5164T\n",
    "        # sample 091 through 180 were run on 5163T\n",
    "        if int(samp_num) <=90:\n",
    "            lane_code = '-5164T/'\n",
    "        else:\n",
    "            lane_code = '-5163T/'\n",
    "        forward_seq_path = seq_path_prefix + samp_num_full + lane_code + '210616ChiA_' + samp_num_full + '_' + '1_sequence.fastq'\n",
    "        reverse_seq_path = seq_path_prefix + samp_num_full + lane_code + '210616ChiA_' + samp_num_full + '_' + '2_sequence.fastq'\n",
    "\n",
    "        manifest_file_list.append({\n",
    "            'sample-id': sample_id,\n",
    "            'forward-absolute-filepath': forward_seq_path,\n",
    "            'reverse-absolute-filepath': reverse_seq_path\n",
    "        })\n",
    "\n",
    "        manifest_df = pd.DataFrame(manifest_file_list)\n",
    "\n",
    "    if save:\n",
    "        manifest_df.to_csv(outdir, sep='\\t', index=False)\n",
    "    return(manifest_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_path = '../data/barcodes_BMC_orig_ds1.xlsx'\n",
    "\n",
    "# will be updated manually (e.g., create a v2 and v3) as needed\n",
    "metadata_df = parse_orig_metadata(metadata_path, outdir='../data/metadata_and_manifest/Blactam_metadata_ds1.tsv', save=True)\n",
    "\n",
    "manifest_df = create_manifest_file(metadata_df, outdir='../data/metadata_and_manifest/Blactam_manifest_ds1.tsv', save=True)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
